gt/main.c

```diff
@@ -810,22 +1024,38 @@ gt_proc(void *arg)
 			 */
 			ret = gt_parse_incoming_pkt(m, &pkt_info);
 			if (ret < 0) {
-				switch (pkt_info.outer_ethertype) {
-				case ETHER_TYPE_IPv4:
-					add_pkt_acl(&acl4, m);
-					continue;
-				case ETHER_TYPE_IPv6:
-					add_pkt_acl(&acl6, m);
+				gt_process_unparsed_incoming_pkt(
+					&acl4, &acl6, &num_arp, arp_bufs,
+					m, pkt_info.outer_ethertype);
+				continue;
+			}
+
+			/*
+			 * If it is a fragmented packet,
+			 * then try to reassemble.
+			 */
+			if (pkt_info.frag) {
+				m = gt_reassemble_incoming_pkt(
+					m, cur_tsc, &pkt_info,
+					&death_row, instance);
+
+				/* Process the death packets. */
+				process_death_row(socket,
+					port, tx_queue, &death_row,
+					instance, gt_conf);
```

Comment:
`process_death_row()` nees one more parameter: `int punish`. When parameter `punish` is false, `process_death_row()` only free the packets in `death_row` without any further processing. When parameter `punish` is true, `process_death_row()` does what it currently does now. The value of parameter `punish` here should be `m == NULL`, that is, we should not punish when a packet was completed because that packet is going through the regular policy.